lib/DxilValidation/DxilContainerValidation.cpp

```diff
@@ -509,14 +506,24 @@ bool VerifySignatureMatches(llvm::Module *pModule, DXIL::SignatureKind SigKind,
 
 static void VerifyPSVMatches(ValidationContext &ValCtx, const void *pPSVData,
                              uint32_t PSVSize) {
+  unsigned ValMajor, ValMinor;
+  ValCtx.DxilMod.GetValidatorVersion(ValMajor, ValMinor);
+  unsigned PSVVersion = hlsl::GetPSVVersion(ValMajor, ValMinor);
   DxilPipelineStateValidation PSV;
-  if (!PSV.InitFromPSV0(pPSVData, PSVSize)) {
+  uint32_t PSVSizeRead = PSVSize;
+  if (!PSV.InitFromPSV0(pPSVData, &PSVSizeRead, PSVVersion)) {
```

Comment:
Before we do this, we should check that `m_uPSVRuntimeInfoSize` (first 4 bytes of data) matches expected size for PSVVersion (`PSVInitInfo(PSVVersion).RuntimeInfoSize()`), reporting error and aborting if not.

Then, there's no need to pass PSVVersion to `InitFromPSV0`, and that overload can be removed/reverted.

Reply:
We should also probably do some additional manual validation of sizes/strides based on the format before diving into the full `InitFromPSV0`:
- `size_t offset = 4`
- start by checking that the `offset + PSVRuntimeInfo_size <= PSVSizeRead`
- map the appropriate PSV structures fitting in size manually to `pPSVData + offset`. `offset += PSVRuntimeInfo_size`
- if ResourceCount > 0:
  - next 4 bytes is PSVResourceBindInfo_size, which must be ` == PSVInitInfo(PSVVersion).ResourceBindInfoSize()`. `offset += 4`
  - check that `offset + PSVResourceBindInfo_size * ResourceCount <= PSVSizeRead`. `offset += PSVResourceBindInfo_size * ResourceCount`
- if `pPSVRuntimeInfo1`:
  - `offset + 4 <= PSVSizeRead`.  Read `uint32_t StringTableSize`. `offset += 4`
    - if `StringTableSize` non-zero:
      - `StringTableSize` must be 4-aligned, `offset + StringTableSize <= PSVSizeRead`, `pPSVData + offset + StringTableSize - 1` must be null char `\0`. `offset += StringTableSize`
    - `offset + 4 <= PSVSizeRead`.  Read `uint32_t SemanticIndexTableEntries`. `offset += 4`
    - `offset + (SemanticIndexTableEntries * 4) <= PSVSizeRead`. `offset += SemanticIndexTableEntries * 4`
  - If `SigInputElements || SigOutputElements || SigPatchConstOrPrimElements`:
    - `offset + 4 <= PSVSizeRead`.  Read `uint32_t PSVSignatureElement_size`. `offset += 4`
    - `PSVSignatureElement_size == PSVInitInfo(PSVVersion).SignatureElementSize()`
    - `offset + PSVSignatureElement_size * (SigInputElements + SigOutputElements  + SigPatchConstOrPrimElements) <= PSVSizeRead`

After this, it's possible to manually go through the ViewID data calculation, but not much you can report other than it being malformed if the overall size isn't correct, since there are no more specific size values in this part of the data that can be independently verified.  Since `InitFromPSV0` will already do that, you can just move on I think.

If anything fails in here, we should not attempt to continue with more checks or load the PSV0 data.

It's a lot, I know, but doing this is necessary to prevent incorrect sizes for records, which can allow either extra unvalidated embedded data, or result in general malformed error when we could be more helpful.

Note how the `ReadOrWrite` method uses the `CheckedReaderWriter` to ease all these calculations and checks.  It would probably be best to write some helper methods similar to that, but with proper error reporting for the validator.  For instance, it's useful to have a method that checks the remaining size before loading some value from the current offset, then increments the offset by the size of that value.  You could simply use the CheckedReaderWriter, but there will be issues with the way errors are handled that might make it hard to report the appropriate error.

Reply:
Working on this.

Reply:
Updated.